src/Kestrel.Core/Internal/Http/HttpParser.cs

```diff
@@ -31,445 +31,281 @@ public HttpParser(bool showErrorDetails)
+                long consumed = reader.Consumed;
 
-                if (done)
+                if (!reader.TryReadToAny(out ReadOnlySpan<byte> headerLine, Eol, advancePastDelimiter: false))
                 {
-                    examined = consumed;
+                    // Couldn't find another delimiter
+                    break;
                 }
-            }
-        }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private unsafe int FindEndOfName(byte* headerLine, int length)
-        {
-            var index = 0;
-            var sawWhitespace = false;
-            for (; index < length; index++)
-            {
-                var ch = headerLine[index];
-                if (ch == ByteColon)
+                int headerLength = headerLine.Length;
+                if (!reader.IsNext(Eol, advancePast: true))
                 {
+                    // Not a good CR/LF pair
+                    RejectCRLF(ref reader, headerLine);
+                    reader.Rewind(reader.Consumed - consumed);
                     break;
                 }
-                if (ch == ByteTab || ch == ByteSpace || ch == ByteCR)
+
+                if (headerLength == 0)
                 {
-                    sawWhitespace = true;
+                    // Consider an empty line to be the end
+                    success = true;
+                    break;
                 }
-            }
 
-            if (index == length || sawWhitespace)
-            {
-                RejectRequestHeader(headerLine, length);
+                TakeSingleHeader(headerLine, handler);
             }
 
-            return index;
+            return success;
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private unsafe void TakeSingleHeader(byte* headerLine, int length, TRequestHandler handler)
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void RejectCRLF(ref BufferReader<byte> reader, ReadOnlySpan<byte> headerLine)
         {
-            // Skip CR, LF from end position
-            var valueEnd = length - 3;
-            var nameEnd = FindEndOfName(headerLine, length);
-
-            // Header name is empty
-            if (nameEnd == 0)
+            if (reader.TryRead(out byte value) && value == ByteCR && reader.End)
             {
-                RejectRequestHeader(headerLine, length);
+                // Incomplete if ends in CR
+                return;
             }
 
-            if (headerLine[valueEnd + 2] != ByteLF)
+            if (headerLine.Length == 0 && value != ByteLF)
             {
-                RejectRequestHeader(headerLine, length);
+                BadHttpRequestException.Throw(RequestRejectionReason.InvalidRequestHeadersNoCRLF);
             }
-            if (headerLine[valueEnd + 1] != ByteCR)
+            else
             {
-                RejectRequestHeader(headerLine, length);
+                RejectRequestHeader(headerLine);
```

Comment:
I think this should change to a message that states that it found an unexpected CR or LF. Then it would be super clear that the detail message is followed by such. My change cuts the span before CR or LF, the previous code cut before any LF. I'd want to do that in a separate change though I think. Thoughts? @pakrym?